In [51]:
#a
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score



df = pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Heart_s.csv")


In [52]:
#b
print (df.keys())

print (df[['Age' , 'Gender', 'MaxHR','AHD']])


Index([u'Age', u'Gender', u'ChestPain', u'RestBP', u'Chol', u'RestECG',
       u'MaxHR', u'Oldpeak', u'Thal', u'AHD'],
      dtype='object')
     Age Gender  MaxHR  AHD
0     63      f    150   No
1     67      f    108  Yes
2     67      f    129  Yes
3     37      f    187   No
4     41      m    172   No
5     56      f    178   No
6     62      m    160  Yes
7     57      m    163   No
8     63      f    147  Yes
9     53      f    155  Yes
10    57      f    148   No
11    56      m    153   No
12    56      f    142  Yes
13    44      f    173   No
14    52      f    162   No
15    57      f    174   No
16    48      f    168  Yes
17    54      f    160   No
18    48      m    139   No
19    49      f    171   No
20    64      f    144   No
21    58      m    162   No
22    58      f    160  Yes
23    58      f    173  Yes
24    60      f    132  Yes
25    50      m    158   No
26    58      m    172   No
27    66      m    114   No
28    43      f    171   No
29    40      f    

In [53]:
#c
feature_cols = ['Age','RestBP', 'Chol', 'RestECG', 'MaxHR','Oldpeak']
X = df[feature_cols]

X[::10]


print(X.shape)

y = df['AHD']

y[::10]

(301, 6)


0       No
10      No
20      No
30      No
40     Yes
50      No
60     Yes
70      No
80      No
90     Yes
100     No
110    Yes
120    Yes
130     No
140    Yes
150     No
160    Yes
170     No
180    Yes
190    Yes
200     No
210    Yes
220     No
230    Yes
240     No
250    Yes
260    Yes
270    Yes
280    Yes
290    Yes
300     No
Name: AHD, dtype: object

In [54]:
#d
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=6)

print(X_train.shape)
print(y_train.shape)

(225, 6)
(225,)


In [55]:
print(X_test.shape)
print(y_test.shape)

(76, 6)
(76,)


In [56]:
#e

k = 3
knn = KNeighborsClassifier(n_neighbors=k) 

knn.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=None, n_neighbors=3, p=2,
           weights='uniform')

Predicting

In [57]:
y_predict = knn.predict(X_test)

print (y_predict)

['Yes' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'No' 'No' 'No' 'No' 'Yes' 'No' 'No'
 'Yes' 'Yes' 'No' 'No' 'No' 'Yes' 'No' 'No' 'Yes' 'No' 'No' 'Yes' 'Yes'
 'Yes' 'No' 'No' 'No' 'Yes' 'No' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'No' 'No'
 'No' 'No' 'No' 'Yes' 'No' 'No' 'No' 'No' 'No' 'Yes' 'No' 'No' 'Yes' 'No'
 'Yes' 'No' 'No' 'No' 'No' 'Yes' 'No' 'Yes' 'No' 'No' 'Yes' 'No' 'No'
 'Yes' 'No' 'No' 'Yes' 'Yes' 'Yes' 'No' 'No']


Accuracy

In [58]:
knn_accuracy = accuracy_score(y_test, y_predict)

print(knn_accuracy)

0.6447368421052632


Decision Tree

In [59]:
my_decisiontree = DecisionTreeClassifier(random_state = 5)

my_decisiontree.fit(X_train, y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=5,
            splitter='best')

Predicting

In [60]:
y_predict = my_decisiontree.predict(X_test)

print (y_predict)

['Yes' 'Yes' 'Yes' 'Yes' 'No' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'Yes' 'No'
 'No' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'No' 'No' 'Yes' 'No' 'No' 'Yes'
 'Yes' 'Yes' 'Yes' 'No' 'Yes' 'No' 'No' 'Yes' 'No' 'No' 'Yes' 'No' 'Yes'
 'Yes' 'Yes' 'Yes' 'No' 'Yes' 'Yes' 'No' 'No' 'Yes' 'Yes' 'No' 'No' 'Yes'
 'No' 'Yes' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'No' 'No' 'No' 'Yes' 'No' 'No'
 'No' 'No' 'Yes' 'No' 'No' 'Yes' 'No' 'Yes' 'No' 'No']


Accuracy 

In [61]:
my_decisiontree_accuracy = accuracy_score(y_test, y_predict)

print(my_decisiontree_accuracy)

0.618421052631579


Logistic Regression

In [62]:
my_logreg = LogisticRegression();

my_logreg.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

Predicting

In [63]:
y_predict = my_logreg.predict(X_test)

print (y_predict)

['Yes' 'Yes' 'No' 'No' 'No' 'Yes' 'Yes' 'Yes' 'No' 'No' 'No' 'Yes' 'No'
 'No' 'Yes' 'Yes' 'No' 'No' 'No' 'Yes' 'No' 'No' 'Yes' 'No' 'No' 'No' 'No'
 'Yes' 'Yes' 'No' 'Yes' 'No' 'No' 'Yes' 'No' 'No' 'Yes' 'No' 'Yes' 'Yes'
 'Yes' 'Yes' 'No' 'Yes' 'Yes' 'No' 'No' 'Yes' 'Yes' 'Yes' 'Yes' 'No' 'No'
 'Yes' 'No' 'No' 'No' 'No' 'No' 'No' 'No' 'No' 'Yes' 'Yes' 'No' 'No' 'No'
 'Yes' 'Yes' 'No' 'Yes' 'Yes' 'No' 'Yes' 'No' 'No']


Accuracy

In [64]:
my_logreg_accuracy = accuracy_score(y_test, y_predict)

print(my_logreg_accuracy)

0.6710526315789473


So far in my prediction for Heart Disease
KNN Accuracy  : 0.6447368421052632 = 64%
Decision Tree Accuracy : 0.618421052631579 = 62%
Logistic Regression Accuracy : 0.6710526315789473 = 67%

The best is Logistic Regression because it resulted to 67%
The one to the be the best is Knn because it resulted to 64%
The worst is Decision tree because it resulted to 62%


In [65]:
#f
def gender_male_to_num(x):
    if x == 'm':
        return 1
    elif x == 'f':
        return 0
def gender_female_to_num(x):
    if x == 'm':
        return 0
    elif x == 'f':
        return 1

In [66]:
df['gender_male_num'] = df['Gender'].apply(gender_male_to_num)
df['gender_female_num'] = df['Gender'].apply(gender_female_to_num)

In [67]:
df[::10]

Age Gender     ChestPain  RestBP  Chol  RestECG  MaxHR  Oldpeak  \
0     63      f       typical     145   233        2    150      2.3   
10    57      f  asymptomatic     140   192        0    148      0.4   
20    64      f       typical     110   211        2    144      1.8   
30    69      m       typical     140   239        0    151      1.8   
40    65      m  asymptomatic     150   225        2    114      1.0   
50    41      m    nontypical     105   198        0    168      0.0   
60    51      m  asymptomatic     130   305        0    142      1.2   
70    65      m    nonanginal     155   269        0    148      0.8   
80    45      f  asymptomatic     104   208        2    148      3.0   
90    62      m  asymptomatic     160   164        2    145      6.2   
100   34      f       typical     118   182        2    174      0.0   
110   56      f  asymptomatic     125   249        2    144      1.2   
120   63      m  asymptomatic     150   407        2    154      4.0   
130   51      f    nonanginal      94   227        0    154      0.0   
140   59      f       typical     170   288        2    159      0.2   
150   42      m  asymptomatic     102   265        2    122      0.6   
160   77      f  asymptomatic     125   304        2    162      0.0   
170   53      f  asymptomatic     142   226        2    111      0.0   
180   56      m  asymptomatic     134   409        2    150      1.9   
190   51      f  asymptomatic     140   298        0    122      4.2   
200   64      m  asymptomatic     180   325        0    154      0.0   
210   38      f       typical     120   231        0    182      3.8   
220   54      m    nonanginal     108   267        2    167      0.0   
230   55      m  asymptomatic     180   327        1    117      3.4   
240   41      m    nontypical     126   306        0    163      0.0   
250   58      f  asymptomatic     146   218        0    105      2.0   
260   58      m    nontypical     136   319        2    152      0.0   
270   46      f  asymptomatic     140   311        0    120      1.8   
280   55      m  asymptomatic     128   205        1    130      2.0   
290   44      f  asymptomatic     120   169        0    144      2.8   
300   38      f    nonanginal     138   175        0    173      0.0   

           Thal  AHD  gender_male_num  gender_female_num  
0         fixed   No                0                  1  
10        fixed   No                0                  1  
20       normal   No                0                  1  
30       normal   No                1                  0  
40   reversable  Yes                1                  0  
50       normal   No                1                  0  
60   reversable  Yes                1                  0  
70       normal   No                1                  0  
80       normal   No                0                  1  
90   reversable  Yes                1                  0  
100      normal   No                0                  1  
110      normal  Yes                0                  1  
120  reversable  Yes                1                  0  
130  reversable   No                0                  1  
140  reversable  Yes                0                  1  
150      normal   No                1                  0  
160      normal  Yes                0                  1  
170  reversable   No                0                  1  
180  reversable  Yes                1                  0  
190  reversable  Yes                0                  1  
200      normal   No                1                  0  
210  reversable  Yes                0                  1  
220      normal   No                1                  0  
230      normal  Yes                1                  0  
240      normal   No                1                  0  
250  reversable  Yes                0                  1  
260      normal  Yes                1                  0  
270  reversable  Yes                0           

Chest Pain

In [68]:
def chestpain_typical_to_numeric(x):
    if x == 'typical':
        return 1
    else :
        return 0

def chestpain_asymptomatic_to_numeric(x):
    if x == 'asymptomatic':
        return 1
    else :
        return 0

def chestpain_nonanginal_to_numeric(x):
    if x == 'nonanginal':
        return 1
    else :
        return 0

def chestpain_nontypical_to_numeric(x):
    if x == 'nontypical':
        return 1
    else :
        return 0

In [69]:
df['chestpain_typical_to_num'] = df['ChestPain'].apply(chestpain_typical_to_numeric)
df['chestpain_asymptomatic_to_num'] = df['ChestPain'].apply(chestpain_asymptomatic_to_numeric)
df['chestpain_nonanginal_to_num'] = df['ChestPain'].apply(chestpain_nonanginal_to_numeric)
df['chestpain_nontypical_to_num'] = df['ChestPain'].apply(chestpain_nontypical_to_numeric)

In [70]:
df[::10]

Age Gender     ChestPain  RestBP  Chol  RestECG  MaxHR  Oldpeak  \
0     63      f       typical     145   233        2    150      2.3   
10    57      f  asymptomatic     140   192        0    148      0.4   
20    64      f       typical     110   211        2    144      1.8   
30    69      m       typical     140   239        0    151      1.8   
40    65      m  asymptomatic     150   225        2    114      1.0   
50    41      m    nontypical     105   198        0    168      0.0   
60    51      m  asymptomatic     130   305        0    142      1.2   
70    65      m    nonanginal     155   269        0    148      0.8   
80    45      f  asymptomatic     104   208        2    148      3.0   
90    62      m  asymptomatic     160   164        2    145      6.2   
100   34      f       typical     118   182        2    174      0.0   
110   56      f  asymptomatic     125   249        2    144      1.2   
120   63      m  asymptomatic     150   407        2    154      4.0   
130   51      f    nonanginal      94   227        0    154      0.0   
140   59      f       typical     170   288        2    159      0.2   
150   42      m  asymptomatic     102   265        2    122      0.6   
160   77      f  asymptomatic     125   304        2    162      0.0   
170   53      f  asymptomatic     142   226        2    111      0.0   
180   56      m  asymptomatic     134   409        2    150      1.9   
190   51      f  asymptomatic     140   298        0    122      4.2   
200   64      m  asymptomatic     180   325        0    154      0.0   
210   38      f       typical     120   231        0    182      3.8   
220   54      m    nonanginal     108   267        2    167      0.0   
230   55      m  asymptomatic     180   327        1    117      3.4   
240   41      m    nontypical     126   306        0    163      0.0   
250   58      f  asymptomatic     146   218        0    105      2.0   
260   58      m    nontypical     136   319        2    152      0.0   
270   46      f  asymptomatic     140   311        0    120      1.8   
280   55      m  asymptomatic     128   205        1    130      2.0   
290   44      f  asymptomatic     120   169        0    144      2.8   
300   38      f    nonanginal     138   175        0    173      0.0   

           Thal  AHD  gender_male_num  gender_female_num  \
0         fixed   No                0                  1   
10        fixed   No                0                  1   
20       normal   No                0                  1   
30       normal   No                1                  0   
40   reversable  Yes                1                  0   
50       normal   No                1                  0   
60   reversable  Yes                1                  0   
70       normal   No                1                  0   
80       normal   No                0                  1   
90   reversable  Yes                1                  0   
100      normal   No                0                  1   
110      normal  Yes                0                  1   
120  reversable  Yes                1                  0   
130  reversable   No                0                  1   
140  reversable  Yes                0                  1   
150      normal   No                1                  0   
160      normal  Yes                0                  1   
170  reversable   No                0                  1   
180  reversable  Yes                1                  0   
190  reversable  Yes                0                  1   
200      normal   No                1                  0   
210  reversable  Yes                0                  1   
220      normal   No                1                  0   
230      normal  Yes                1                  0   
240      normal   No                1                  0   
250  reversable  Yes                0                  1   
260      normal  Yes                1                  0   
270  reversable  Yes

Thal

In [71]:
def thal_fixed_to_numeric(x):
    if x == 'fixed':
        return 1
    else :
        return 0

def thal_normal_to_numeric(x):
    if x == 'normal':
        return 1
    else :
        return 0

def thal_reversable_to_numeric(x):
    if x == 'reversable':
        return 1
    else :
        return 0

In [72]:
df['thal_fixed_to_num'] = df['Thal'].apply(thal_fixed_to_numeric)
df['thal_normal_to_num'] = df['Thal'].apply(thal_normal_to_numeric)
df['thal_reversable_to_num'] = df['Thal'].apply(thal_reversable_to_numeric)

In [73]:
df[::10]

Age Gender     ChestPain  RestBP  Chol  RestECG  MaxHR  Oldpeak  \
0     63      f       typical     145   233        2    150      2.3   
10    57      f  asymptomatic     140   192        0    148      0.4   
20    64      f       typical     110   211        2    144      1.8   
30    69      m       typical     140   239        0    151      1.8   
40    65      m  asymptomatic     150   225        2    114      1.0   
50    41      m    nontypical     105   198        0    168      0.0   
60    51      m  asymptomatic     130   305        0    142      1.2   
70    65      m    nonanginal     155   269        0    148      0.8   
80    45      f  asymptomatic     104   208        2    148      3.0   
90    62      m  asymptomatic     160   164        2    145      6.2   
100   34      f       typical     118   182        2    174      0.0   
110   56      f  asymptomatic     125   249        2    144      1.2   
120   63      m  asymptomatic     150   407        2    154      4.0   
130   51      f    nonanginal      94   227        0    154      0.0   
140   59      f       typical     170   288        2    159      0.2   
150   42      m  asymptomatic     102   265        2    122      0.6   
160   77      f  asymptomatic     125   304        2    162      0.0   
170   53      f  asymptomatic     142   226        2    111      0.0   
180   56      m  asymptomatic     134   409        2    150      1.9   
190   51      f  asymptomatic     140   298        0    122      4.2   
200   64      m  asymptomatic     180   325        0    154      0.0   
210   38      f       typical     120   231        0    182      3.8   
220   54      m    nonanginal     108   267        2    167      0.0   
230   55      m  asymptomatic     180   327        1    117      3.4   
240   41      m    nontypical     126   306        0    163      0.0   
250   58      f  asymptomatic     146   218        0    105      2.0   
260   58      m    nontypical     136   319        2    152      0.0   
270   46      f  asymptomatic     140   311        0    120      1.8   
280   55      m  asymptomatic     128   205        1    130      2.0   
290   44      f  asymptomatic     120   169        0    144      2.8   
300   38      f    nonanginal     138   175        0    173      0.0   

           Thal  AHD  gender_male_num  gender_female_num  \
0         fixed   No                0                  1   
10        fixed   No                0                  1   
20       normal   No                0                  1   
30       normal   No                1                  0   
40   reversable  Yes                1                  0   
50       normal   No                1                  0   
60   reversable  Yes                1                  0   
70       normal   No                1                  0   
80       normal   No                0                  1   
90   reversable  Yes                1                  0   
100      normal   No                0                  1   
110      normal  Yes                0                  1   
120  reversable  Yes                1                  0   
130  reversable   No                0                  1   
140  reversable  Yes                0                  1   
150      normal   No                1                  0   
160      normal  Yes                0                  1   
170  reversable   No                0                  1   
180  reversable  Yes                1                  0   
190  reversable  Yes                0                  1   
200      normal   No                1                  0   
210  reversable  Yes                0                  1   
220      normal   No                1                  0   
230      normal  Yes                1                  0   
240      normal   No                1                  0   
250  reversable  Yes                0                  1   
260      normal  Yes                1                  0   
270  reversable  Yes

In [74]:
#g
feature_cols = [
    'Age','RestBP', 'Chol', 'RestECG', 'MaxHR','Oldpeak',
    'gender_male_num','gender_female_num',
    'chestpain_typical_to_num','chestpain_asymptomatic_to_num','chestpain_nonanginal_to_num','chestpain_nontypical_to_num',
    'thal_fixed_to_num','thal_normal_to_num','thal_reversable_to_num']
X = df[feature_cols]
X[::10]

Age  RestBP  Chol  RestECG  MaxHR  Oldpeak  gender_male_num  \
0     63     145   233        2    150      2.3                0   
10    57     140   192        0    148      0.4                0   
20    64     110   211        2    144      1.8                0   
30    69     140   239        0    151      1.8                1   
40    65     150   225        2    114      1.0                1   
50    41     105   198        0    168      0.0                1   
60    51     130   305        0    142      1.2                1   
70    65     155   269        0    148      0.8                1   
80    45     104   208        2    148      3.0                0   
90    62     160   164        2    145      6.2                1   
100   34     118   182        2    174      0.0                0   
110   56     125   249        2    144      1.2                0   
120   63     150   407        2    154      4.0                1   
130   51      94   227        0    154      0.0                0   
140   59     170   288        2    159      0.2                0   
150   42     102   265        2    122      0.6                1   
160   77     125   304        2    162      0.0                0   
170   53     142   226        2    111      0.0                0   
180   56     134   409        2    150      1.9                1   
190   51     140   298        0    122      4.2                0   
200   64     180   325        0    154      0.0                1   
210   38     120   231        0    182      3.8                0   
220   54     108   267        2    167      0.0                1   
230   55     180   327        1    117      3.4                1   
240   41     126   306        0    163      0.0                1   
250   58     146   218        0    105      2.0                0   
260   58     136   319        2    152      0.0                1   
270   46     140   311        0    120      1.8                0   
280   55     128   205        1    130      2.0                1   
290   44     120   169        0    144      2.8                0   
300   38     138   175        0    173      0.0                0   

     gender_female_num  chestpain_typical_to_num  \
0                    1                         1   
10                   1                         0   
20                   1                         1   
30                   0                         1   
40                   0                         0   
50                   0                         0   
60                   0                         0   
70                   0                         0   
80                   1                         0   
90                   0                         0   
100                  1                         1   
110                  1                         0   
120                  0                         0   
130                  1                         0   
140                  1                         1   
150                  0                         0   
160                  1                         0   
170                  1                         0   
180                  0                         0   
190                  1                         0   
200                  0                         0   
210                  1                         1   
220                  0                         0   
230                  0                         0   
240                  0                         0   
250                  1                         0   
260                  0                         0   
270                  1                         0   
280                  0                         0   
290                  1                         0   
300                  1                         0   

     chestpain_asymptomatic_to_num  chestpain_nonanginal_to_num  \
0                                0                            0   
10                           

In [75]:
print(X.shape)

(301, 15)


In [76]:
y = df['AHD']

y[::10]

0       No
10      No
20      No
30      No
40     Yes
50      No
60     Yes
70      No
80      No
90     Yes
100     No
110    Yes
120    Yes
130     No
140    Yes
150     No
160    Yes
170     No
180    Yes
190    Yes
200     No
210    Yes
220     No
230    Yes
240     No
250    Yes
260    Yes
270    Yes
280    Yes
290    Yes
300     No
Name: AHD, dtype: object

In [77]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, 
                                                    random_state=6)

In [78]:
print(X_train.shape)
print(y_train.shape)

(225, 15)
(225,)


In [79]:
print(X_test.shape)
print(y_test.shape)

(76, 15)
(76,)


KNN

In [80]:
k=3
knn = KNeighborsClassifier(n_neighbors=k) 

knn.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=None, n_neighbors=3, p=2,
           weights='uniform')

Predicting

In [81]:
y_predict = knn.predict(X_test)

print (y_predict)

['Yes' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'No' 'No' 'No' 'No' 'Yes' 'No' 'No'
 'Yes' 'Yes' 'No' 'No' 'No' 'Yes' 'No' 'No' 'Yes' 'No' 'No' 'Yes' 'Yes'
 'Yes' 'No' 'No' 'No' 'Yes' 'No' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'No' 'No'
 'No' 'No' 'No' 'Yes' 'No' 'No' 'No' 'No' 'No' 'Yes' 'No' 'No' 'Yes' 'No'
 'Yes' 'No' 'No' 'No' 'No' 'Yes' 'No' 'Yes' 'No' 'No' 'Yes' 'No' 'No'
 'Yes' 'No' 'No' 'Yes' 'Yes' 'Yes' 'No' 'No']


Accuracy

In [82]:
knn_accuracy = accuracy_score(y_test, y_predict)

print(knn_accuracy)

0.6447368421052632


Decision Tree

In [83]:
my_decisiontree = DecisionTreeClassifier(random_state = 5)

my_decisiontree.fit(X_train, y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=5,
            splitter='best')

Predicting

In [84]:
y_predict = my_decisiontree.predict(X_test)

print (y_predict)

['Yes' 'Yes' 'No' 'No' 'No' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'Yes' 'No'
 'No' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'No' 'No' 'Yes' 'No' 'Yes' 'Yes'
 'Yes' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'No' 'No' 'No' 'Yes' 'No' 'No'
 'Yes' 'No' 'No' 'No' 'Yes' 'Yes' 'Yes' 'No' 'Yes' 'Yes' 'Yes' 'Yes' 'No'
 'No' 'Yes' 'No' 'Yes' 'Yes' 'No' 'No' 'No' 'Yes' 'No' 'Yes' 'Yes' 'No'
 'No' 'No' 'Yes' 'Yes' 'No' 'No' 'Yes' 'No' 'Yes' 'Yes' 'Yes']


Accuracy

In [85]:
my_decisiontree_accuracy = accuracy_score(y_test, y_predict)

print(my_decisiontree_accuracy)

0.7105263157894737


Logistic Regression

In [86]:
my_logreg= LogisticRegression();


my_logreg.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

Predicting

In [87]:
y_predict = my_logreg.predict(X_test)

print (y_predict)

['Yes' 'Yes' 'No' 'No' 'No' 'Yes' 'Yes' 'Yes' 'No' 'No' 'No' 'Yes' 'No'
 'No' 'Yes' 'Yes' 'No' 'No' 'No' 'Yes' 'No' 'No' 'Yes' 'No' 'No' 'Yes'
 'No' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'No' 'No' 'No' 'Yes' 'No' 'No' 'No'
 'No' 'Yes' 'No' 'Yes' 'Yes' 'Yes' 'No' 'Yes' 'No' 'Yes' 'Yes' 'No' 'No'
 'Yes' 'No' 'No' 'Yes' 'No' 'No' 'No' 'Yes' 'No' 'Yes' 'Yes' 'No' 'No'
 'No' 'Yes' 'Yes' 'No' 'Yes' 'Yes' 'Yes' 'Yes' 'No' 'No']


Accuracy

In [88]:
my_logreg_accuracy = accuracy_score(y_test, y_predict)

print(my_logreg_accuracy)


0.7631578947368421


In [89]:
#h

feature_cols = [
    'Age','RestBP', 'Chol', 'RestECG', 'MaxHR','Oldpeak',
    'gender_male_num','gender_female_num',
    'chestpain_typical_to_num','chestpain_asymptomatic_to_num','chestpain_nonanginal_to_num','chestpain_nontypical_to_num',
    'thal_fixed_to_num','thal_normal_to_num','thal_reversable_to_num']
X = df[feature_cols]
X[::10]

Age  RestBP  Chol  RestECG  MaxHR  Oldpeak  gender_male_num  \
0     63     145   233        2    150      2.3                0   
10    57     140   192        0    148      0.4                0   
20    64     110   211        2    144      1.8                0   
30    69     140   239        0    151      1.8                1   
40    65     150   225        2    114      1.0                1   
50    41     105   198        0    168      0.0                1   
60    51     130   305        0    142      1.2                1   
70    65     155   269        0    148      0.8                1   
80    45     104   208        2    148      3.0                0   
90    62     160   164        2    145      6.2                1   
100   34     118   182        2    174      0.0                0   
110   56     125   249        2    144      1.2                0   
120   63     150   407        2    154      4.0                1   
130   51      94   227        0    154      0.0                0   
140   59     170   288        2    159      0.2                0   
150   42     102   265        2    122      0.6                1   
160   77     125   304        2    162      0.0                0   
170   53     142   226        2    111      0.0                0   
180   56     134   409        2    150      1.9                1   
190   51     140   298        0    122      4.2                0   
200   64     180   325        0    154      0.0                1   
210   38     120   231        0    182      3.8                0   
220   54     108   267        2    167      0.0                1   
230   55     180   327        1    117      3.4                1   
240   41     126   306        0    163      0.0                1   
250   58     146   218        0    105      2.0                0   
260   58     136   319        2    152      0.0                1   
270   46     140   311        0    120      1.8                0   
280   55     128   205        1    130      2.0                1   
290   44     120   169        0    144      2.8                0   
300   38     138   175        0    173      0.0                0   

     gender_female_num  chestpain_typical_to_num  \
0                    1                         1   
10                   1                         0   
20                   1                         1   
30                   0                         1   
40                   0                         0   
50                   0                         0   
60                   0                         0   
70                   0                         0   
80                   1                         0   
90                   0                         0   
100                  1                         1   
110                  1                         0   
120                  0                         0   
130                  1                         0   
140                  1                         1   
150                  0                         0   
160                  1                         0   
170                  1                         0   
180                  0                         0   
190                  1                         0   
200                  0                         0   
210                  1                         1   
220                  0                         0   
230                  0                         0   
240                  0                         0   
250                  1                         0   
260                  0                         0   
270                  1                         0   
280                  0                         0   
290                  1                         0   
300                  1                         0   

     chestpain_asymptomatic_to_num  chestpain_nonanginal_to_num  \
0                                0                            0   
10                           

In [90]:
print(X.shape)

(301, 15)


In [91]:
y = df['AHD']

y[::10]

0       No
10      No
20      No
30      No
40     Yes
50      No
60     Yes
70      No
80      No
90     Yes
100     No
110    Yes
120    Yes
130     No
140    Yes
150     No
160    Yes
170     No
180    Yes
190    Yes
200     No
210    Yes
220     No
230    Yes
240     No
250    Yes
260    Yes
270    Yes
280    Yes
290    Yes
300     No
Name: AHD, dtype: object

KNN

In [92]:
k = 3
knn = KNeighborsClassifier(n_neighbors=k) 


knn_accuracy_list = cross_val_score(knn, X, y, cv=10, scoring='accuracy')

In [93]:
print (knn_accuracy_list)

[0.70967742 0.64516129 0.51612903 0.66666667 0.63333333 0.46666667
 0.66666667 0.66666667 0.55172414 0.72413793]


In [94]:
knn_accuracy_cv = knn_accuracy_list.mean()

print(knn_accuracy_cv)

0.6246829810901001


Decision Tree

In [95]:
my_decisiontree = DecisionTreeClassifier(random_state = 5)

my_decisiontree_accuracy_list = cross_val_score(my_decisiontree, X, y, cv=10, scoring='accuracy')

In [96]:
print (my_decisiontree_accuracy_list)

[0.74193548 0.70967742 0.77419355 0.8        0.8        0.7
 0.6        0.6        0.65517241 0.82758621]


In [97]:
my_decisiontree_cv = my_decisiontree_accuracy_list.mean()

print(my_decisiontree_cv)

0.7208565072302558


Logistic Regression

In [98]:
my_logreg = LogisticRegression();


my_logreg_accuracy_list = cross_val_score(my_logreg, X, y, cv=10, scoring='accuracy')

In [99]:
print (my_logreg_accuracy_list)

[0.77419355 0.80645161 0.83870968 0.93333333 0.9        0.66666667
 0.8        0.83333333 0.79310345 0.79310345]


In [100]:
my_logreg_cv = my_logreg_accuracy_list.mean()

print(my_logreg_cv)

0.8138895068594735


Final Prediction for Cross Validation 

Knn Prediction Accuracy 0.6246829810901001 = 62%
Decision Tree Accuracy  : 0.7208565072302558 = 72%
Logistic Regression : 0.8138895068594735 = 81%

Best : Logistic Regression -> Increased a lot more from past prediction in part f
Next Best : Decision Tree ->  Increased from past prediction in Part f
Worst : KNN ->  Decreased from past prediction in part f


The best option seems to be Cross Validation instead Splitting the Data as we did at first,
and to use Linear Regression for the best prediction. 